src/main/java/nl/ordina/jobcrawler/service/VacancyService.java

```diff
@@ -1,58 +1,98 @@
+        Set<Skill> existingSkills = skillService.linkToExistingSkills(vacancy.getSkills());
+        existingSkills.forEach(skill -> skill.addVacancy(vacancy)); // add the vacancy to the skill
+        vacancy.setSkills(existingSkills); // add the skills to the vacancy
+
+        //saving the vacancy if it has a correct URL
+        if (vacancy.checkURL()) return vacancyRepository.saveAndFlush(vacancy);
+        else return null;
     }
 
-    /* Get */
-    public List<Vacancy> getAllJobs() {
-        return repository.findAll();
+    //******** Getting ********//
+    public List<Vacancy> getAllVacancies() {
+        return vacancyRepository.findAll();
     }
 
-    public List<Vacancy> getJobsWithSkill(String skill) {
-        return repository.findAll()
-                .stream()
-                .filter(a -> a.getSkillSet().toString().toLowerCase().contains(skill.toLowerCase()))
-                .collect(Collectors.toList());
+    public Optional<Vacancy> getVacancyByID(UUID id) {
+        return vacancyRepository.findById(id);
     }
 
-    public Optional<Vacancy> getByID(UUID id) {
-        return repository.findById(id);
+    public Set<Vacancy> getVacanciesBySkill(String skill) {
+        return skillService.getVacanciesBySkill(skill);
     }
 
-    public Optional<Vacancy> getExistingRecord(String url) {
-        return repository.findByVacancyURLEquals(url);
+    public List<Vacancy> getVacanciesByBroker(String broker) {
+        return vacancyRepository.findByBrokerEquals(broker);
     }
 
-    public List<Vacancy> getJobsByBroker(String broker) {
-        return repository.findByBrokerEquals(broker);
+    public Optional<Vacancy> getExistingVacancy(String url) {
+        return vacancyRepository.findByVacancyURLEquals(url);
     }
 
-    /* Delete */
+
+    //******** Deleting ********//
+    // Deletes vacancy and accompanying skills and relations from the database //
     public void delete(UUID id) {
-        repository.deleteById(id);
+        Vacancy vacancyToDelete = vacancyRepository.findById(id).orElseThrow(() -> new VacancyNotFoundException(id));
+        Set<Skill> skillsToDelete = new HashSet<Skill>(vacancyToDelete.getSkills()); // saves the skills of the vacancy
+
+        skillsToDelete.forEach(skill -> skill.removeVacancy(vacancyToDelete)); // remove all vacancy relations
+
+        vacancyRepository.deleteById(id); //deletes the vacancy and all relations to the skills
+
+        skillService.deleteSkillsIfNoRelations(skillsToDelete); //deletes the skills if there are no more relations pointing to it
+    }
+
+
+    //******** Updating ********//
+    // Updates Vacancy with a given new vacancy (can write to all 3 tables)//
+    public Vacancy replace(UUID id, Vacancy newJob) {
+        return vacancyRepository.findById(id)
+                .map(job -> {
+                    System.out.println("** Job before modification: \n" + job);
+
+                    job.setVacancyURL(newJob.getVacancyURL());
```

Comment:
The reason to use setters here is because you need to mutate this object, because this object is tracked by the entity manager. If you create a new object (for example, with Lombok's toBuilder), that new object is not tracked, and not updated by default in the database. I think there are ways around that problem, but this should work.

Reply:
Needs a big knowledge dive from me, so I made a jira subtask of it (NEX-74) and will resolve the comment for now.